rust2vec/src/storage.rs

```diff
@@ -47,39 +49,127 @@ pub type CowArray1<'a, A> = CowArray<'a, A, Ix1>;
-        match self {
-            Storage::NdArray(ref data) => CowArray::Borrowed(data.index_axis(Axis(0), idx)),
-        }
+        CowArray::Owned(self.view().row(idx).to_owned())
     }
 
     pub fn view(&self) -> ArrayView2<f32> {
         match self {
             Storage::NdArray(ref data) => data.view(),
+            Storage::Mmap { map, shape } => unsafe {
+                ArrayView2::from_shape_ptr(*shape, map.as_ptr() as *const f32)
+            },
+        }
+    }
+
+    fn write_ndarray_chunk<S, W>(data: ArrayBase<S, Ix2>, mut write: W) -> Result<(), Error>
+    where
+        S: Data<Elem = f32>,
+        W: Write + Seek,
+    {
+        // n_rows: 8 bytes, n_cols: 4 bytes, type_id: 4, matrix
+        let chunk_len = 16 + (data.rows() * data.cols() * mem::size_of::<f32>());
+
+        write.write_u32::<LittleEndian>(ChunkIdentifier::NdArray as u32)?;
+        write.write_u64::<LittleEndian>(chunk_len as u64)?;
+        write.write_u64::<LittleEndian>(data.rows() as u64)?;
+        write.write_u32::<LittleEndian>(data.cols() as u32)?;
+        write.write_u32::<LittleEndian>(f32::type_id())?;
+
+        // Write padding, such that the embedding matrix starts on at
+        // a multiple of the size of f32 (4 bytes). This is necessary
+        // for memory mapping a matrix. Interpreting the raw u8 data
+        // as a proper f32 array requires that the data is aligned in
+        // memory. However, we cannot always memory map the starting
+        // offset of the matrix directly, since mmap(2) requires a
+        // file offset that is page-aligned. Since the page size is
+        // always a larger power of 2 (e.g. 2^12), which is divisable
+        // by 4, the offset of the matrix with regards to the page
+        // boundary is also a multiple of 4.
+
+        let n_padding = padding::<f32>(write.seek(SeekFrom::Current(0))?);
+        let padding = vec![0; n_padding as usize];
+        write.write(&padding)?;
+
+        for row in data.outer_iter() {
+            for col in row.iter() {
+                write.write_f32::<LittleEndian>(*col)?;
+            }
         }
+
+        Ok(())
+    }
+}
+
+impl MmapChunk for Storage {
+    fn mmap_chunk(read: &mut BufReader<File>) -> Result<Self, Error> {
+        ensure!(
+            read.read_u32::<LittleEndian>()? == ChunkIdentifier::NdArray as u32,
+            "invalid chunk identifier for NdArray"
+        );
+
+        // Read and discard chunk length.
+        read.read_u64::<LittleEndian>()?;
+
+        let rows = read.read_u64::<LittleEndian>()? as usize;
+        let cols = read.read_u32::<LittleEndian>()? as usize;
+        let shape = Ix2(rows, cols);
+
+        ensure!(
+            read.read_u32::<LittleEndian>()? == f32::type_id(),
+            "Expected single precision floating point matrix for NdArray."
+        );
+
+        let n_padding = padding::<f32>(read.seek(SeekFrom::Current(0))?);
+        read.seek(SeekFrom::Current(n_padding as i64))?;
+
```

Comment:
Same idea as above, just verifying I'm understanding the logic here:
`read.seek(SeekFrom::Current(0))?)` returns the position of the reader after reading the known parts. In theory the `seek` call could be replaced by adding up the width of the previously read items, although that would make maintenance more bug-prone, if the header were ever subject to change.

Reply:
In theory yes, but we would also have to include the counts of previous chunks and the header.